likelihood.py

```diff
@@ -0,0 +1,232 @@
+        cmsymr0[~f] = self.rmin**(-self.etamR[~f])* cmr0[arr_i[~f] - Nmax//2]
+
+        cmsymr0[-1] = cmsymr0[-1] / 2
+        cmsymr0[0] = cmsymr0[0] / 2
+
+        P0t = np.real(cmsymr0*self.J0k_arr).sum(axis=1)
+        P0int = interpolate.InterpolatedUnivariateSpline(self.kbins3,P0t)(self.k)
+        return P0int
+
+    def loglkl(self, params, box=384, npart=128, timesteps=10):
+        """Compute the log-likelihood of the model, given the data and covariance loaded in __init__.
+        The cosmology and model parameters are put in the params list.
+        The fastpm accuracy parameters box, npart and timesteps are kw args.
+        Model parameters:
+        0 - h the hubble parameter
+        1 - Omega_0 the total matter density
+        2 - scalar_amp the primordial amplitude of fluctuations at k = 0.05 Mpc^{-1}
+        3 - The linear bias, a variable which encodes how much galaxies differ from cold dark matter.
+        """
+        #Cosmology parameters
+        hubble = params[0]
+        omega0 = params[1]
+        scalar_amp = params[2]
+        #ns is not measured by BOSS very well, so fixed
+        #ns = params[3]
+        linear_bias = params[3]
+
+        ## COMPUTE SPECTRA
+        # Run fastPM for power spectrum with cosmological parameters.
+        sim = SimulationICs(hubble = hubble,
+                            omega0 = omega0,
+                            scalar_amp = scalar_amp,
+                            ns = 0.9649,
+                            fastpm_bin = "fastpm",
+                            box = box,
+                            npart = npart,
+                            timesteps = timesteps,
+                            redend = self.z,
+                            )
+        sim.make_simulation()
+        powerspec = sim.powerspecs[-1]
+        kk = sim.kk[-1]
+        #Galaxy power spectrum with multipoles, for reference.
+        #Powergal = lambda mu, l: (linear_bias + ff * mu**2)**2 * sim.powerspecs[-1]
+        #For higher moments we should do a multipole integral. Here just use l = 0 for simplicity.
+        #Neglect the redshift space distortions because they are degenerate with linear bias
+        #for the monopole.
+        powergal = linear_bias**2 * powerspec
+        powgalint = interpolate.InterpolatedUnivariateSpline(kk,powergal)
+
+        # Now do a Fourier-transform to include the window function
+        factor = np.exp(-1.*(self.kbins3*hubble/2.)**4.)*self.tmp_factor
+        factint = interpolate.InterpolatedUnivariateSpline(self.kbins3*hubble, factor)
+        Pdisc = powgalint(self.k)*factint(self.k)
+
+        # Now compute chi^2: this is:
+        #(Pth W - Pexp)^T C^-1 (Pth W - Pexp)
+        chi2 = np.inner(Pdisc[:,0],np.inner(self.invcovWW,Pdisc[:,0]))
+        chi2 += np.inner(self.Pk0,np.inner(self.invcov,self.Pk0))
+        chi2 += -2.*np.inner(Pdisc[:,0],np.inner(self.invcovW,self.Pk0))
```

Comment:
This line also won't work because `np.inner(self.invcovW,self.Pk0)` is (128, ) and Pdisc is (48, )

`self.invcovW` also needs to be interpolated onto `self.k`

Reply:
Now Pdisc is interpolated to kbins3 not self.k and the shapes should match